src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -9,227 +9,316 @@ internal struct FileStatus
+
+        // We track intent of creation to know whether or not we want to (1) create a
+        // DirectoryInfo around this status struct or (2) actually are part of a DirectoryInfo.
+        internal bool InitiallyDirectory { get; set; }
+
+        internal bool IsMainCacheValid => _initializedMainCache == 0;
+
+        internal bool IsSecondaryCacheValid => _initializedSecondaryCache == 0;
+
+        private bool IsValid =>
+            _initializedMainCache == 0 && // Should always be successfully refreshed
+            (_initializedSecondaryCache == -1 || _initializedSecondaryCache == 0); // Only refreshed when path is detected to be a symbolic link
+
+        internal void EnsureCachesInitialized(ReadOnlySpan<char> path, bool continueOnError = false)
+        {
+            if (!IsValid)
             {
-                // User belongs to a group that effectively owns the file
-                readBit = Interop.Sys.Permissions.S_IRGRP;
-                writeBit = Interop.Sys.Permissions.S_IWGRP;
+                RefreshCaches(path);
             }
-            else
+
+            if (!continueOnError)
             {
-                // Others permissions
-                readBit = Interop.Sys.Permissions.S_IROTH;
-                writeBit = Interop.Sys.Permissions.S_IWOTH;
+                ThrowOnCacheInitializationError(path);
             }
-#endif
-
-            return ((_fileStatus.Mode & (int)readBit) != 0 && // has read permission
-                (_fileStatus.Mode & (int)writeBit) == 0);     // but not write permission
         }
 
-        public FileAttributes GetAttributes(ReadOnlySpan<char> path, ReadOnlySpan<char> fileName)
+        internal FileAttributes GetAttributes(ReadOnlySpan<char> path, ReadOnlySpan<char> fileName)
         {
             // IMPORTANT: Attribute logic must match the logic in FileSystemEntry
 
-            EnsureStatInitialized(path);
+            EnsureCachesInitialized(path);
 
             if (!_exists)
+            {
                 return (FileAttributes)(-1);
+            }
 
             FileAttributes attributes = default;
 
-            if (IsReadOnly(path))
+            if (HasReadOnlyFlag)
+            {
                 attributes |= FileAttributes.ReadOnly;
+            }
 
-            if ((_fileStatus.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFLNK)
+            if (HasSymbolicLinkFlag)
+            {
                 attributes |= FileAttributes.ReparsePoint;
+            }
 
             if (_isDirectory)
+            {
                 attributes |= FileAttributes.Directory;
+            }
 
             // If the filename starts with a period or has UF_HIDDEN flag set, it's hidden.
-            if (fileName.Length > 0 && (fileName[0] == '.' || (_fileStatus.UserFlags & (uint)Interop.Sys.UserFlags.UF_HIDDEN) == (uint)Interop.Sys.UserFlags.UF_HIDDEN))
+            if (fileName.Length > 0 && (fileName[0] == '.' || HasHiddenFlag))
```

Comment:
nit: this is not the first place I can see the logic for `.` prefix detection. Could it be moved to a helper method and reused everywhere?

Reply:
The only other place is in `FileSystemEntry.HasHiddenPrefix` (different struct). I could centralize the logic in `FileStatus`.